src/main/java/bio/terra/workspace/service/resource/reference/ReferenceResourceService.java

```diff
@@ -0,0 +1,138 @@
+  private final SamService samService;
+  private final WorkspaceService workspaceService;
+
+  @Autowired
+  public ReferenceResourceService(
+      JobService jobService,
+      ResourceDao resourceDao,
+      SamService samService,
+      WorkspaceService workspaceService) {
+    this.jobService = jobService;
+    this.resourceDao = resourceDao;
+    this.samService = samService;
+    this.workspaceService = workspaceService;
+  }
+
+  @Traced
+  public ReferenceResource createReferenceResource(
+      ReferenceResource resource, AuthenticatedUserRequest userReq) {
+    // TODO: Fix this based on resolution of permission model issue.
+    workspaceService.validateWorkspaceAndAction(
+        userReq, resource.getWorkspaceId(), SamConstants.SAM_WORKSPACE_WRITE_ACTION);
+    resource.validate();
+
+    String jobDescription =
+        String.format(
+            "Create reference %s; id %s; name %s",
+            resource.getResourceType(), resource.getResourceId(), resource.getName());
+    JobBuilder createJob =
+        jobService
+            .newJob(
+                jobDescription,
+                UUID.randomUUID().toString(),
+                CreateReferenceResourceFlight.class,
+                resource,
+                userReq)
+            .addParameter(
+                WorkspaceFlightMapKeys.ResourceKeys.RESOURCE_TYPE, resource.getResourceType());
+
+    UUID resourceIdResult = createJob.submitAndWait(UUID.class);
+    if (!resourceIdResult.equals(resource.getResourceId())) {
+      throw new InvalidMetadataException("Input and output resource ids do not match");
+    }
+
+    return getReferenceResource(resource.getWorkspaceId(), resourceIdResult, userReq);
+  }
+
+  /**
+   * At this moment, the only updates on references we are doing are to name and description so this
+   * is a common implementation. If we get more complicated, we can break it out.
+   *
+   * @param workspaceId workspace of interest
+   * @param resourceId resource to update
+   * @param name name to change - may be null
+   * @param description description to change - may be null
+   */
+  public void updateReferenceResource(
+      UUID workspaceId,
+      UUID resourceId,
+      String name,
+      String description,
+      AuthenticatedUserRequest userReq) {
+    workspaceService.validateWorkspaceAndAction(
+        userReq, workspaceId, SamConstants.SAM_WORKSPACE_WRITE_ACTION);
+    resourceDao.updateReferenceResource(workspaceId, resourceId, name, description);
+  }
+
+  /**
+   * Delete a reference. The only state we hold for a reference is in the metadata database so we
+   * directly delete that.
+   *
+   * @param workspaceId workspace of interest
+   * @param resourceId resource to delete
+   * @param userReq authenticated user
+   */
+  public void deleteReferenceResource(
+      UUID workspaceId, UUID resourceId, AuthenticatedUserRequest userReq) {
+    // TODO: Fix this based on resolution of permission model issue.
```

Comment:
is there a ticket for this issue?

Reply:
Not yet. We are discussing on Thursday in the WSM sync meeting with the AS team. I have a list of about 7 items that I will make into tickets based on how this work comes out.